framework/DataObjects/DataSet.py

```diff
@@ -0,0 +1,1853 @@
+          if index > numInData + numInCollector - 1:
+            self.raiseAnError(IndexError,'Requested index "{}" but only have {} entries (zero-indexed)!'.format(index,numInData+numInCollector))
+          ## otherwise, take from the collector
+          else:
+            rlz = self._getRealizationFromCollectorByIndex(index - numInData)
+        ## otherwise, take from the data
+        else:
+          rlz = self._getRealizationFromDataByIndex(index, unpackXArray)
+      # handle "-" requests (counting from the end): first end of collector, or if not then end of data
+      else:
+        # caller is requesting so many "from the end", so work backwards
+        ## if going back further than what's in the collector ...
+        if abs(index) > numInCollector:
+          ## if further going back further than what's in the data, then we don't have that entry
+          if abs(index) > numInData + numInCollector:
+            self.raiseAnError(IndexError,'Requested index "{}" but only have {} entries!'.format(index,numInData+numInCollector))
+          ## otherwise, grab the requested index from the data
+          else:
+            rlz = self._getRealizationFromDataByIndex(index + numInCollector, unpackXArray)
+        ## otherwise, grab the entry from the collector
+        else:
+          rlz = self._getRealizationFromCollectorByIndex(index)
+      return rlz
+    ## END select by index
+    ## START collect by matching realization
+    else: # matchDict must not be None
+      # if nothing in data, try collector
+      if numInData == 0:
+        # if nothing in data OR collector, we can't have a match
+        if numInCollector == 0:
+          return 0,None
+        # otherwise, get it from the collector
+        else:
+          index,rlz = self._getRealizationFromCollectorByValue(matchDict,tol=tol)
+      # otherwise, first try to find it in the data
+      else:
+        index,rlz = self._getRealizationFromDataByValue(matchDict,tol=tol, unpackXArray=unpackXArray)
+        # if no match found in data, try in the collector (if there's anything in it)
+        if rlz is None:
+          if numInCollector > 0:
+            index,rlz = self._getRealizationFromCollectorByValue(matchDict,tol=tol)
+      return index,rlz
+
+  def remove(self,realization=None,variable=None):
+    """
+      Used to remove either a realization or a variable from this data object.
+      @ In, realization, dict or int, optional, (matching or index of) realization to remove
+      @ In, variable, str, optional, name of "column" to remove
+      @ Out, None
+    """
+    if self.size == 0:
+      self.raiseAWarning('Called "remove" on DataObject, but it is empty!')
+      return
+    noData = self._data is None or len(self._data) == 0
+    noColl = self._collector is None or len(self._collector) == 0
+    assert(not (realization is None and variable is None))
+    assert(not (realization is not None and variable is not None))
+    assert(self._data is not None)
```

Comment:
These sure look like runtime conditions, which means they should be raised exceptions, not assertions.

Reply:
These assertions no long apply, so I removed them.